### PR TITLE
[REVIEW] Add Series.dropna()

### LINKED ISSUE #911: [FEA] Implement `dropna` method.
**Is your feature request related to a problem? Please describe.**
A common use case for pandas users is to drop null rows. In writing pyjanitor, I also wrote helper method that dropped all empty rows and columns, building on top of the `dropna` API. 

**Describe the solution you'd like**
I hope to see `dropna()` implemented for cuDF.

**Describe alternatives you've considered**
As this is a request to implement a pandas API method call, I have not considered other alternatives.

**Additional context**
No additional context needed, I think, though I'm happy to provide more if the devs need it! :smile:

### PR DESCRIPTION
**Pull request by shwina**
This PR adds the dropna() methods to Series ~and DataFrame~.
 
Partially addresses #911. Supersedes #1126 and depends on #1622.

DataFrame.dropna() will be added after #1744.

### CODE REVIEW

File: python/cudf/dataframe/string.py [hunk 1]
```diff
@@ -558,13 +560,44 @@ def fillna(self, fill_value, inplace=False):
...
         """
         Get unique strings in the data
         """
         result = StringColumn(self.nvcategory.keys())
         return result
 
+    def normalize_binop_value(self, other):
+        if isinstance(other, column.Column):
+            return other.astype(self.dtype)
+        elif hasattr(other, '__str__'):
```
Comment (beckernick) [inline python/cudf/dataframe/string.py]: Since integers/floats have the __str__ attribute, I believe we would end up with something like this returning True:

```python
s = cudf.Series(['0', '1', '2'])
s == 2
False
False
True
```

This is unexpected behavior for this coming from pandas, in which all of these would be False.

```python
s = pd.Series(['0', '1', '2'])
s == 2
0    False
1    False
2    False
dtype: bool
```

Comment (shwina) [reply python/cudf/dataframe/string.py]: Ah thanks. So then should we just check if other is a `str`? Does that cover the common use cases? The only other value I can think of handling would be `None`

Comment (beckernick) [reply python/cudf/dataframe/string.py]: That would likely cover the common use cases. I believe that users generally expect to be able to compare a string column to

- A single string (e.g., `'word'` or `'two words'`)
- Null representation (e.g., `None`, or `np.NaN` in pandas)
- A comparison array of the same length as the original array (in which case the broadcasting is unnecessary), perhaps such as a `column` (to use your example) but almost certainly also a `Series`.

Our normalization process should be designed so it can handle users supplying at least these three core types of comparisons.

Note that the behavior of comparing to None/np.NaN in pandas is interesting.

```python
s = pd.Series(['a', 'b', np.nan, 'c', None])
print(s)
print(s == None)
print(s == np.nan)
0       a
1       b
2     NaN
3       c
4    None
dtype: object
0    False
1    False
2    False
3    False
4    False
dtype: bool
0    False
1    False
2    False
3    False
4    False
dtype: bool
```

Finally, we should think carefully about what kinds of arrays we want to allow users to use as the comparison array. As of now, passing a `column` as the comparison array with Series backed by NumericalColumns will fail. We may want to be consistent with that behavior for StringColumns, or update it across the board to allow directly comparing against `columns`. Additionally, whether or not we intend to allow users to pass a CPU array such as `['a', 'b', 'c']` is a broader discussion that I think we can push down the road. @kkraus14's opinion on all of this would be valuable as well.

Comment (shwina) [reply python/cudf/dataframe/string.py]: Thanks for the insights

**Comment by raydouglass** [comment]
rerun tests

**Comment by thomcom** [comment]
Can you update the description to include some of the design decisions you had to make when creating this PR? There's a lot going on in the list of changes, why so much just to add `dropna()`? Thanks.

**Comment by shwina** [comment]
@thomcom It appeared so because I didn't rebase after #1622 was merged. Should be looking better now but still very much a WIP

File: cpp/src/stream_compaction/copy_if.cuh [hunk 2]
```diff
@@ -352,6 +353,14 @@ gdf_column copy_if(gdf_column const &input, Filter filter,
                           output, input, block_offsets, filter,
                           input.valid != nullptr, stream);
 
+    // TODO: TEMPORARY FIX
```
Comment (harrism) [inline cpp/src/stream_compaction/copy_if.cuh]: I'm working on a PR for this, including adding tests for stream compaction on cudf::nvstring_category columns.

Comment (shwina) [reply cpp/src/stream_compaction/copy_if.cuh]: Great! This is only a placeholder change

File: python/cudf/bindings/stream_compaction.pxd [hunk 3]
```diff
@@ -9,7 +9,4 @@ from cudf.bindings.cudf_cpp cimport *
 
 cdef extern from "stream_compaction.hpp" namespace "cudf" nogil:
 
-    cdef gdf_column apply_boolean_mask(const gdf_column &input,
```
[the remark this answers is not in the document]
Comment (shwina) [reply python/cudf/bindings/stream_compaction.pxd]: Sorry - not sure what happened here. Will restore

File: python/cudf/dataframe/string.py (hunk 1 above)
Comment (thomcom) [reply python/cudf/dataframe/string.py]: Just throwing in that I don't think direct `column` comparison needs to be supported - Users should be using `Dataframes` and `Series` and not obligated to deal with any deeper support classes.

File: python/cudf/dataframe/series.py [hunk 4]
```diff
@@ -700,6 +700,16 @@ def masked_assign(self, value, mask):
...
         return self._copy_construct(data=data)
 
+    def dropna(self):
+        """
+        Return a Series with null values removed.
+        """
+        if self.null_count == 0:
+            return self
+        data = self._column.dropna()
+        index = self.index.loc[~self.isna()]
```
Comment (shwina) [inline python/cudf/dataframe/series.py]: This is inefficient but we'll have to wait for the multi-column dropna to do better

File: cpp/src/stream_compaction/copy_if.cuh [hunk 5]
```diff
@@ -353,10 +353,9 @@ gdf_column copy_if(gdf_column const &input, Filter filter,
                           scatter_functor<Filter, block_size, per_thread>{},
                           output, input, block_offsets, filter,
                           input.valid != nullptr, stream);
-
     CHECK_STREAM(stream);
   }
-  
+
```
Comment (thomcom) [inline cpp/src/stream_compaction/copy_if.cuh]: Just a comment - this is a no-op file modification that is orthogonal to the work you're actually doing. Obviously the change is "safe", but it clutters up the changelog filelist for this commit and could be substantially confusing downstream.

Comment (shwina) [reply cpp/src/stream_compaction/copy_if.cuh]: Agreed. I'll drop this change in a follow-up commit.

File: python/cudf/bindings/stream_compaction.pxd (hunk 3 above)
[the remark this answers is not in the document]
Comment (shwina) [reply python/cudf/bindings/stream_compaction.pxd]: Undone

File: cpp/src/stream_compaction/copy_if.cuh (hunk 2 above)
Comment (shwina) [reply cpp/src/stream_compaction/copy_if.cuh]: Removed

File: cpp/src/stream_compaction/copy_if.cuh (hunk 5 above)
Comment (shwina) [reply cpp/src/stream_compaction/copy_if.cuh]: Done
